File: src/main/java/me/zeroeightsix/kami/command/commands/ConfigCommand.java
package me.zeroeightsix.kami.command.commands;

import com.mojang.brigadier.CommandDispatcher;
import com.mojang.brigadier.arguments.StringArgumentType;
import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.builder.RequiredArgumentBuilder;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.command.Command;
import me.zeroeightsix.kami.command.KamiCommandSource;
import net.minecraft.server.command.CommandSource;
import net.minecraft.text.LiteralText;

import static me.zeroeightsix.kami.util.Texts.*;
import static net.minecraft.util.Formatting.*;

/**
 * Created by 086 on 14/10/2018.
 */
public class ConfigCommand extends Command {

    private static final DynamicCommandExceptionType FAILED_EXCEPTION = new DynamicCommandExceptionType(o -> new LiteralText(o.toString()));

    @Override
    public void register(CommandDispatcher<CommandSource> dispatcher) {
        dispatcher.register(
            LiteralArgumentBuilder.<CommandSource>literal("config")
                    .then(LiteralArgumentBuilder.<CommandSource>literal("reload")
                            .executes(context -> {
                                KamiMod.loadConfiguration();
                                ((KamiCommandSource) context.getSource()).sendFeedback(
                                        f(GOLD, append(
                                                lit("Reloaded configuration "),
                                                flit(YELLOW, KamiMod.getConfigName()),
                                                lit("!")
                                        ))
                                );
                                return 0;
                            })
                    )
                    .then(LiteralArgumentBuilder.<CommandSource>literal("save")
                            .executes(context -> {
                                KamiMod.saveConfiguration();
                                ((KamiCommandSource) context.getSource()).sendFeedback(
                                        f(GOLD, append(
                                                lit("Saved configuration "),
                                                flit(YELLOW, KamiMod.getConfigName()),
                                                lit("!")
                                        ))
                                );
                                return 0;
                            })
                    )
                    .then(LiteralArgumentBuilder.<CommandSource>literal("switch")
                            .then(RequiredArgumentBuilder.<CommandSource, String>argument("filename", StringArgumentType.string())
                                .executes(context -> {
                                    String filename = context.getArgument("filename", String.class);
                                    if (!KamiMod.isFilenameValid(filename)) {
                                        throw FAILED_EXCEPTION.create("Invalid filename '" + filename + "'!");
                                    }
                                    KamiCommandSource source = (KamiCommandSource) context.getSource();
                                    KamiMod.saveConfiguration();
                                    source.sendFeedback(f(GOLD, append(
                                            lit("Saved "),
                                            flit(YELLOW, KamiMod.getConfigName())
                                    )));
                                    KamiMod.setLastConfigName(filename);
                                    source.sendFeedback(f(GOLD, append(
                                            lit("Set "),
                                            flit(YELLOW, "KAMILastConfig.txt"),
                                            lit(" to "),
                                            flit(LIGHT_PURPLE, filename)
                                    )));
                                    KamiMod.loadConfiguration();
                                    source.sendFeedback(f(GOLD, append(
                                            lit("Loaded "),
                                            flit(YELLOW, filename),
                                            lit("!")
                                    )));
                                    return 0;
                                })
                            )
                    )
        );
    }

}

File: src/main/kotlin/me/zeroeightsix/kami/command/commands/ConfigCommand.kt
package me.zeroeightsix.kami.command.commands

import com.mojang.brigadier.CommandDispatcher
import com.mojang.brigadier.arguments.StringArgumentType
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.mojang.brigadier.builder.RequiredArgumentBuilder
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.command.Command
import me.zeroeightsix.kami.command.KamiCommandSource
import me.zeroeightsix.kami.util.Texts
import net.minecraft.server.command.CommandSource
import net.minecraft.text.LiteralText
import net.minecraft.util.Formatting
import java.util.function.Function

/**
 * Created by 086 on 14/10/2018.
 */
object ConfigCommand : Command() {
    private val FAILED_EXCEPTION =
        DynamicCommandExceptionType(Function { o: Any ->
            LiteralText(o.toString())
        })

    override fun register(dispatcher: CommandDispatcher<CommandSource>) {
        dispatcher.register(
            LiteralArgumentBuilder.literal<CommandSource>("config")
                .then(
                    LiteralArgumentBuilder.literal<CommandSource>("reload")
                        .executes { context: CommandContext<CommandSource> ->
                            KamiMod.loadConfiguration()
                            (context.source as KamiCommandSource).sendFeedback(
                                Texts.f(
                                    Formatting.GOLD, Texts.append(
                                        Texts.lit("Reloaded configuration "),
                                        Texts.flit(
                                            Formatting.YELLOW,
                                            KamiMod.getConfigName()
                                        ),
                                        Texts.lit("!")
                                    )
                                )
                            )
                            0
                        }
                )
                .then(
                    LiteralArgumentBuilder.literal<CommandSource>("save")
                        .executes { context: CommandContext<CommandSource> ->
                            KamiMod.saveConfiguration()
                            (context.source as KamiCommandSource).sendFeedback(
                                Texts.f(
                                    Formatting.GOLD, Texts.append(
                                        Texts.lit("Saved configuration "),
                                        Texts.flit(
                                            Formatting.YELLOW,
                                            KamiMod.getConfigName()
                                        ),
                                        Texts.lit("!")
                                    )
                                )
                            )
                            0
                        }
                )
                .then(
                    LiteralArgumentBuilder.literal<CommandSource>("switch")
                        .then(
                            RequiredArgumentBuilder.argument<CommandSource, String>(
                                "filename",
                                StringArgumentType.string()
                            )
                                .executes { context: CommandContext<CommandSource> ->
                                    val filename =
                                        context.getArgument("filename", String::class.java)
                                    if (!KamiMod.isFilenameValid(filename)) {
                                        throw FAILED_EXCEPTION.create("Invalid filename '$filename'!")
                                    }
                                    val source = context.source as KamiCommandSource
                                    KamiMod.saveConfiguration()
                                    source.sendFeedback(
                                        Texts.f(
                                            Formatting.GOLD, Texts.append(
                                                Texts.lit("Saved "),
                                                Texts.flit(
                                                    Formatting.YELLOW,
                                                    KamiMod.getConfigName()
                                                )
                                            )
                                        )
                                    )
                                    KamiMod.setLastConfigName(filename)
                                    source.sendFeedback(
                                        Texts.f(
                                            Formatting.GOLD, Texts.append(
                                                Texts.lit("Set "),
                                                Texts.flit(
                                                    Formatting.YELLOW,
                                                    "KAMILastConfig.txt"
                                                ),
                                                Texts.lit(" to "),
                                                Texts.flit(
                                                    Formatting.LIGHT_PURPLE,
                                                    filename
                                                )
                                            )
                                        )
                                    )
                                    KamiMod.loadConfiguration()
                                    source.sendFeedback(
                                        Texts.f(
                                            Formatting.GOLD, Texts.append(
                                                Texts.lit("Loaded "),
                                                Texts.flit(
                                                    Formatting.YELLOW,
                                                    filename
                                                ),
                                                Texts.lit("!")
                                            )
                                        )
                                    )
                                    0
                                }
                        )
                )
        )
    }
}